### ISSUE TITLE
`az functionapp keys list` throws 'Bad Request'

### ISSUE THREAD

**Issue by kimjamia** [Azure/azure-cli#20892]
## Describe the bug

**Command Name**
`az functionapp keys list`

**Errors:**
```
Operation returned an invalid status 'Bad Request'
```

## To Reproduce:
- Linux Function App v4.
- Configured to pull container image from Azure Container Registry using managed identity.
- `az functionapp keys list -g {} -n {}`

## Expected Behavior
- Lists the keys (or shows an empty list like I would've expected in this particular case) or
- more informative error message.

## Environment Summary
```
Linux-4.4.0-19041-Microsoft-x86_64-with-debian-bullseye-sid, Ubuntu 20.04 LTS
Python 3.6.10
Installer: DEB

azure-cli 2.31.0

Extensions:
aks-preview 0.5.49
```

**Comment by yonzhan** [comment]
route to service team

**Comment by kimjamia** [comment]
I updated the steps to reproduce. This is most likely related to the fact that the storage account was created with `--public-network-access Disabled`. However, when trying to create a minimal setup that reproduces this issue, I couldn't get it to throw Bad Request. Instead it returns:
```
(NotFound) Error retrieving master key. <html>
<head><title>404 Not Found</title></head>
<body>
<center><h1>404 Not Found</h1></center>
<hr><center>nginx/1.21.5</center>
</body>
</html>

Code: NotFound
Message: Error retrieving master key. <html>
<head><title>404 Not Found</title></head>
<body>
<center><h1>404 Not Found</h1></center>
<hr><center>nginx/1.21.5</center>
</body>
</html>
```

**Comment by Crossbow78** [comment]
Our deployment pipeline retrieves the Function App master key to invoke a HTTP trigger as part of an automated smoke-test after deployment of our application. We notice that this step sometimes randomly fails with a "Bad Request", and when we try to retrieve the key a few minutes later it succeeds.

This is the command:
`az functionapp keys list -g "xxxxx" -n "xxxxx" --query "masterKey" --output tsv`

This is the error reported by 'az':
`ERROR: Operation returned an invalid status 'Bad Request'`

And the Activity Log on the Function App contains an entry similar to this:
```json
{
    "eventName": {
        "value": "EndRequest",
        "localizedValue": "End request"
    },
    "level": "Error",
    "operationName": {
        "value": "Microsoft.Web/sites/host/listkeys/action",
        "localizedValue": "List Web Apps Functions Host Keys"
    },
    "resourceId": "/subscriptions/xxxxxxxx/resourceGroups/xxxxxxxx/providers/Microsoft.Web/sites/xxxxxxxx/host/default",
    "status": {
        "value": "Failed",
        "localizedValue": "Failed"
    },
    "subStatus": {
        "value": "BadRequest",
        "localizedValue": "Bad Request (HTTP Status Code: 400)"
    },
    "properties": {
        "statusCode": "BadRequest",
        "serviceRequestId": null,
        "statusMessage": "{\"Code\":\"BadRequest\",\"Message\":\"Encountered an error (InternalServerError) from host runtime.\",\"Target\":null,\"Details\":[{\"Message\":\"Encountered an error (InternalServerError) from host runtime.\"},{\"Code\":\"BadRequest\"},{\"ErrorEntity\":{\"Code\":\"BadRequest\",\"Message\":\"Encountered an error (InternalServerError) from host runtime.\"}}],\"Innererror\":null}",
        "eventCategory": "Administrative",
        "message": "Microsoft.Web/sites/host/listkeys/action",
    }
}
```

I'm not sure what the previously mentioned `--public-network-access` option does and I don't think it's related to our issue, but for completeness sake I'll mention that our storage account is created via an ARM template incremental deployment with the following properties:
```
    "properties": {
        "allowBlobPublicAccess": false,
        "minimumTlsVersion": "TLS1_2",
        "supportsHttpsTrafficOnly": true,
        "networkAcls": {
            "bypass": "AzureServices",
            "virtualNetworkRules": [
                {
                    "id": "[resourceId('Microsoft.Network/virtualNetworks/subnets', variables('vnet-name'), variables('subnet-streamprocessor-name'))]",
                    "action": "Allow"
                }
            ],
            "ipRules": [],
            "defaultAction": "Deny"
        }
    },
```

**Comment by DanoThom** [comment]
Can we get an update on this please. I've been dealing with this issue for several months now and it seems very hit or miss as to whether this az cli request works or not. If it was broken then I'd be more understanding but the variability of the request succeeding is most frustrating. Also, I've tried with both public network access and allow blob public access enabled and I still get this issue:

![image](https://user-images.githubusercontent.com/4973075/156166138-233f1061-76b4-4d39-b821-60bf7d7b4942.png)


![image](https://user-images.githubusercontent.com/4973075/156166026-7c001ae9-d578-4909-b12b-1b7600a6540f.png)

**Comment by DanoThom** [comment]
FYI, this is failing in the Azure Portal also for certain functions that have function key interactions:
<img width="731" alt="image" src="https://user-images.githubusercontent.com/4973075/157030563-ac5ff317-d0a2-43d3-8b51-b30e4bb10c92.png">

**Comment by vats28** [comment]
Look like no update from microsoft, same issue with windows function app v4

**Comment by panchagnula** [comment]
hey all - it would be best to open a support case for these - since this is not specific to CLI. I did try to report these to the function apps team directly, but not something we can take action via CLI.

**Comment by anirudhgarg** [comment]
Yes would be best to open a support ticket or you can transfer this issue to the Functions Host repo: https://github.com/Azure/azure-functions-host/issues

**Comment by panchagnula** [comment]
route to CXP to help here.. Thanks!

**Comment by navba-MSFT** [comment]
Az CLi was onboarded to CXP from Feb 1st. Removing the CXP Attention label from this issue considering this issue was created in Jan. 
CC @yonzhan

**Comment by MrMeterPing** [comment]
Is there any update?
Steps to reproduce:

> az storage account create --name $(StorageAccountName) --resource-group $(StorageResourceGroupName)
> 
> az appservice plan create --name $(AppServicePlanName) --resource-group $(AppServicePlanResourceGroupName) --sku $(AppServicePlanSku) --location $(AppServicePlanLocation)
> 
> az functionapp create --resource-group $(FunctionResourceGroupName) --plan $(AppServicePlanPath) --name $(FunctionName) --storage-account $(StorageAccountPath) --functions-version $(FunctionVersion) --os-type $(FunctionOs) --runtime dotnet --disable-app-insights true --app-insights-key $(ApplicationInsightsImbasKey) --subnet $(FunctionSubnetPath)
> 
> az functionapp keys list --name $(FunctionName) --resource-group $(FunctionResourceGroupName)


Last command returns: Operation returned an invalid status 'Bad Request'

az rest command returns:

> az rest --method post --uri "/subscriptions/xyz/resourceGroups/rg-func/providers/Microsoft.Web/sites/func-test/host/default/listKeys?api-version=2022-03-01" --query functionKeys.default --output tsv

> Bad Request({"Code":"BadRequest","Message":"Encountered an error (InternalServerError) from host runtime.","Target":null,"Details":[{"Message":"Encountered an error (InternalServerError) from host runtime."},{"Code":"BadRequest"},{"ErrorEntity":{"Code":"BadRequest","Message":"Encountered an error (InternalServerError) from host runtime."}}],"Innererror":null})

Also in the Azure Portal the App Key are not shown and cannot be set:
![image](https://user-images.githubusercontent.com/101572840/166706232-744b7183-87b2-4e44-81de-fbda6b7a0328.png)

Do I miss something how to create a function using azure cli? How can I add a key to my function?

**Comment by panchagnula** [comment]
> Is there any update? Steps to reproduce:
> 
> > az storage account create --name $(StorageAccountName) --resource-group $(StorageResourceGroupName)
> > az appservice plan create --name $(AppServicePlanName) --resource-group $(AppServicePlanResourceGroupName) --sku $(AppServicePlanSku) --location $(AppServicePlanLocation)
> > az functionapp create --resource-group $(FunctionResourceGroupName) --plan $(AppServicePlanPath) --name $(FunctionName) --storage-account $(StorageAccountPath) --functions-version $(FunctionVersion) --os-type $(FunctionOs) --runtime dotnet --disable-app-insights true --app-insights-key $(ApplicationInsightsImbasKey) --subnet $(FunctionSubnetPath)
> > az functionapp keys list --name $(FunctionName) --resource-group $(FunctionResourceGroupName)
> 
> Last command returns: Operation returned an invalid status 'Bad Request'
> 
> az rest command returns:
> 
> > az rest --method post --uri "/subscriptions/xyz/resourceGroups/rg-func/providers/Microsoft.Web/sites/func-test/host/default/listKeys?api-version=2022-03-01" --query functionKeys.default --output tsv
> 
> > Bad Request({"Code":"BadRequest","Message":"Encountered an error (InternalServerError) from host runtime.","Target":null,"Details":[{"Message":"Encountered an error (InternalServerError) from host runtime."},{"Code":"BadRequest"},{"ErrorEntity":{"Code":"BadRequest","Message":"Encountered an error (InternalServerError) from host runtime."}}],"Innererror":null})
> 
> Also in the Azure Portal the App Key are not shown and cannot be set: ![image](https://user-images.githubusercontent.com/101572840/166706232-744b7183-87b2-4e44-81de-fbda6b7a0328.png)
> 
> Do I miss something how to create a function using azure cli? How can I add a key to my function?

@MrMeterPing unfortunately, since this is not a CLI client specific problem but rather API/runtime please use the link above to report the issue https://github.com/Azure/azure-functions-host/issues & some one from functions team can help move this forward. //CC: @anirudhgarg
